reference-implementation/lib/abstract-ops/ecmascript.js

```diff
@@ -34,3 +34,67 @@ exports.TransferArrayBuffer = O => {
 exports.IsDetachedBuffer = O => {
   return isFakeDetached in O;
 };
+
+exports.Call = (F, V, args) => {
+  if (typeof F !== 'function') {
+    throw new TypeError('Argument is not a function');
+  }
+
+  return Function.prototype.apply.call(F, V, args);
+};
+
+exports.GetMethod = (V, P) => {
+  const func = V[P];
+  if (func === undefined || func === null) {
+    return undefined;
+  }
+  if (typeof func !== 'function') {
+    throw new TypeError();
+  }
+  return func;
+};
+
+exports.GetIterator = (obj, hint = 'sync', method) => {
+  assert(hint === 'sync' || hint === 'async');
+  if (method === undefined) {
+    if (hint === 'async') {
+      method = exports.GetMethod(obj, Symbol.asyncIterator);
+      if (method === undefined) {
+        const syncMethod = exports.GetMethod(obj, Symbol.iterator);
+        const syncIterator = exports.GetIterator(obj, 'sync', syncMethod);
+        return syncIterator; // TODO sync to async iterator
```

Comment:
I made an implementation for [`CreateAsyncFromSyncIterator`](https://tc39.es/ecma262/#sec-createasyncfromsynciterator)... but I don't know how I feel about it. 😅
```javascript
function CreateAsyncFromSyncIterator(iteratorRecord) {
  const syncIterator = {
    [Symbol.iterator]: () => iteratorRecord.iterator
  };
  const asyncIterator = (async function* () {
    return yield* syncIterator;
  }());
  const nextMethod = asyncIterator.next;
  return { iterator: asyncIterator, nextMethod, done: false };
}
```
Basically, I use an async generator function that delegates to the given sync iterator (with `yield*`), and use that as the async iterator.

It *works*, but maybe it's too confusing or distracting for the *reference* implementation? 🤷 We could also leave it as it is now, with a comment explaining that the reference implementation doesn't really care if `GetIterator()` returns a sync or an async iterator.

Reply:
That's amazing. I feel like I can just barely understand it, but it's a stretch.

I have a slight preference for leaving it as is, on the grounds that a comment saying "we don't need to do this bit" is easier to understand than the actual implementation. But on the other hand, we do try to keep as close to the standard text as possible, so it could go either way. I will let @domenic cast the deciding vote on this one.

Reply:
> The ReadableStream.from() implementation already wraps all results from next() and return() in new promises, so we don't really need the sync-to-async adapter for that.

Hmm, interesting. Maybe it shouldn't? But, [`for-await-of`](https://tc39.es/ecma262/#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset) does, indirectly, by using Await().

But what is the point of async-from-sync iterator then? I know I introduced it for a reason back when I wrote that spec...

I'll ask around.

Reply:
It looks like AsyncFromSyncIterator is kind of pointless: https://github.com/tc39/ecma262/issues/1765

So yeah, maybe stating that, with a link to that issue thread, is best.

Reply:
No, wait. The issue is what happens if you have a sync iterator of promises. With AsyncFromSyncIterator, you get the promises unwrapped. Without it, you get the promises.

So I think omitting this step would break test cases like `ReadableStream.from([promiseFor1, promiseFor2])`. That should be equivalent to `ReadableStream.from([1, 2])` but without the AsyncFromSyncIterator, it would give you a stream where `(await reader.read()).value` was a promise.

Reply:
I added a test for the general case of [a sync iterable of promises](https://github.com/web-platform-tests/wpt/blob/9ce2d37efeb10b9b6b8df22ddd0ca33b0af936a7/streams/readable-streams/from.any.js#L52-L64), and the specific case of [an array of promises](https://github.com/web-platform-tests/wpt/blob/9ce2d37efeb10b9b6b8df22ddd0ca33b0af936a7/streams/readable-streams/from.any.js#L9-L14). These tests require a correct implementation of `CreateAsyncFromSyncIterator`, so I added the above snippet (with extra comments) to the reference implementation. 🙂